### PR TITLE
Integration of Atlantic Service and HTTP Client Infrastructure

### PR DESCRIPTION
**Pull request by Mohiiit**
## Summary
This PR introduces the Atlantic service integration along with reusable HTTP client infrastructure components.

## Changes
- Integrated Atlantic service for proving capabilities
- Implemented a generic HTTP client for reusability across services
- Created a request builder utility for standardized request construction

## Technical Details
- The generic HTTP client provides helper functions that can be reused for future HTTP client implementations
- Request builder utility simplifies and standardizes the process of creating HTTP requests
- Atlantic service integration leverages the new HTTP infrastructure

## Testing
Test also added for the atlantic service

### CODE REVIEW

File: .env.test [hunk 1]
```diff
@@ -53,18 +53,29 @@ DATABASE_NAME="orchestrator"
...
-SHARP_PROOF_LAYOUT="small"
+SNOS_PROOF_LAYOUT="dynamic"
```
Comment (apoorvsadana) [inline .env.test:65]: repeated

Comment (apoorvsadana) [reply .env.test:65]: fix snos issue with layout as you mentioned

File: crates/orchestrator/src/config.rs [hunk 2]
```diff
@@ -115,6 +119,18 @@ pub async fn init_config() -> color_eyre::Result<Arc<Config>> {
     // us stop using the generic omniqueue abstractions for message ack/nack
     let queue = build_queue_client();
 
+    let snos_proof_layout = match settings_provider.get_settings_or_panic("SNOS_PROOF_LAYOUT").as_str() {
+        "all_cairo" => {
```
Comment (apoorvsadana) [inline crates/orchestrator/src/config.rs:123]: this will also become dynamic

File: crates/orchestrator/src/tests/config.rs [hunk 3]
```diff
@@ -234,6 +235,7 @@ impl TestConfigBuilder {
             queue,
             storage,
             alerts,
+            LayoutName::all_cairo,
```
Comment (apoorvsadana) [inline crates/orchestrator/src/tests/config.rs:238]: dynamic

File: crates/prover-clients/atlantic-service/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,54 @@
+[package]
+name = "atlantic-service"
+version.workspace = true
+edition.workspace = true
+
+[dependencies]
+alloy.workspace = true
+async-trait.workspace = true
+base64 = "0.22.1"
```
Comment (apoorvsadana) [inline crates/prover-clients/atlantic-service/Cargo.toml:9]: move to main cargo.toml

File: crates/prover-clients/atlantic-service/src/client.rs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            SettlementLayer::Starknet => Box::new(StarknetLayer),
+        };
+
+        Self { client, proving_layer }
+    }
+
+    pub async fn add_job(
+        &self,
+        pie_file: &Path,
+        _proof_layout: LayoutName,
```
Comment (apoorvsadana) [inline crates/prover-clients/atlantic-service/src/client.rs:68]: match proof_layout {
LayoutName::dynamic => "dynamic", // why
_ => proof_layout.to_str()
}

File: crates/prover-clients/atlantic-service/tests/lib.rs
Comment (apoorvsadana) [inline crates/prover-clients/atlantic-service/tests/lib.rs:1]: let's confirm with herodotus if we're ok doing this given its mock fast hash true

File: crates/prover-clients/sharp-service/tests/constants.rs [hunk 6]
```diff
@@ -1,2 +1,2 @@
-pub const CAIRO_PIE_PATH: &str = "/tests/artifacts/238996-SN.zip";
+pub const CAIRO_PIE_PATH: &str = "/tests/artifacts/fibonacci.zip";
```
Comment (apoorvsadana) [inline crates/prover-clients/sharp-service/tests/constants.rs:1]: let's revert

File: crates/utils/src/http_client.rs [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+
+    /// Adds a query parameter to the request.
+    pub fn query_param(mut self, key: &str, value: &str) -> Self {
+        self.query_params.insert(key.to_string(), value.to_string());
+        self
+    }
+
+    /// Sets the request body.
+    /// Note: Currently assumes string data. For binary data, this would need to be modified.
+    pub fn body(mut self, body: &str) -> Self {
```
Comment (apoorvsadana) [inline crates/utils/src/http_client.rs:282]: body<T: Serialize>(mut self, body: T) -> Self {
serde_json::to_string/to_bytes
}

File: crates/utils/src/http_client.rs
Comment (apoorvsadana) [inline crates/utils/src/http_client.rs:1]: add test cases to this file

File: docker-compose.yml
Comment (apoorvsadana) [inline docker-compose.yml:1]: check in #madara, let's delete this

File: .env.test [hunk 8]
```diff
@@ -53,18 +53,30 @@ DATABASE_NAME="orchestrator"
...
-SHARP_PROOF_LAYOUT="small"
+SNOS_LAYOUT_NAME="all_cairo"
+
```
Comment (apoorvsadana) [inline .env.test:65]:
```suggestion
```

File: crates/orchestrator/src/config.rs [hunk 9]
```diff
@@ -115,6 +121,30 @@ pub async fn init_config() -> color_eyre::Result<Arc<Config>> {
...
 
+    let snos_layout_name = match settings_provider.get_settings_or_panic("SNOS_LAYOUT_NAME").as_str() {
+        "all_cairo" => {
+            log::warn!(
+                "Using all_cairo layout for SNOS. This is probably not provable and so it's not recommended for \
+                 production use."
+            );
+            LayoutName::all_cairo
+        }
+        "starknet_with_keccak" => LayoutName::starknet_with_keccak,
```
Comment (apoorvsadana) [inline crates/orchestrator/src/config.rs:132]:
```suggestion
```

File: crates/utils/src/http_client.rs [hunk 10]
```diff
@@ -0,0 +1,845 @@
...
+//! with default configurations. It supports:
+//! - Default headers, query parameters, and form data
+//! - Client-level TLS certificates and identities
+//! - Request-level customization
+//! - Body handling (currently optimized for string payloads)
+//!
+//! # Body Handling
+//! The current implementation assumes bodies are UTF-8 string data. When merging default body
+//! parameters with request bodies, they are concatenated with '&' as a separator. For binary
+//! data or different formats, the implementation would need to be modified.
```
Comment (apoorvsadana) [inline crates/utils/src/http_client.rs:13]: would need to change this

File: crates/utils/src/http_client.rs [hunk 11]
```diff
@@ -0,0 +1,845 @@
...
+        assert_eq!(client.default_query_params.get("key1").unwrap(), "value1");
+        assert_eq!(client.default_query_params.get("key2").unwrap(), "value 2");
+    }
+
+    /// Tests body parameter handling:
+    /// - Parameters are properly concatenated with '&'
+    /// - Empty parameters are handled correctly
+    /// - URL encoding is applied appropriately
+    #[test]
+    fn test_builder_default_body_params() {
```
Comment (apoorvsadana) [inline crates/utils/src/http_client.rs:399]: i don't think default body param is neede

File: crates/utils/src/http_client.rs [hunk 12]
```diff
@@ -0,0 +1,845 @@
...
+    /// - Non-existent file handling
+    #[test]
+    fn test_multipart_form_file() {
+        let mock_server = httpmock::MockServer::start();
+
+        let client = HttpClient::builder(&mock_server.base_url()).build().unwrap();
+
+        let file_path: PathBuf = "../orchestrator/src/tests/artifacts/fibonacci.zip".parse().unwrap();
+
+        let request = client.request().form_file("file", &file_path, "fibonacci.zip");
```
Comment (apoorvsadana) [inline crates/utils/src/http_client.rs:542]: assert none above

File: crates/utils/src/http_client.rs [hunk 13]
```diff
@@ -0,0 +1,845 @@
...
+    }
+
+    /// # Mock Tests
+
+    /// Tests successful response scenarios using mock server:
+    /// - Different response codes
+    /// - Various content types
+    /// - Response header handling
+    #[tokio::test]
+    async fn test_mock_successful_response() {
```
Comment (apoorvsadana) [inline crates/utils/src/http_client.rs:637]: we can delete this

File: crates/utils/src/http_client.rs [hunk 14]
```diff
@@ -0,0 +1,845 @@
...
+        mock.assert();
+    }
+
+    /// Validates error response handling:
+    /// - Different HTTP error codes
+    /// - Timeout scenarios
+    /// - Network errors
+    /// - Malformed responses
+    #[tokio::test]
+    async fn test_mock_error_responses() {
```
Comment (apoorvsadana) [inline crates/utils/src/http_client.rs:665]: can be removed

File: crates/utils/src/http_client.rs [hunk 15]
```diff
@@ -0,0 +1,845 @@
...
+    }
+
+    /// # Error Handling Tests
+
+    /// Tests TLS certificate and identity handling:
+    /// - Valid certificate addition
+    /// - Invalid certificate handling
+    /// - Identity verification
+    #[test]
+    fn test_certificate_handling() {
```
Comment (apoorvsadana) [inline crates/utils/src/http_client.rs:716]: technically this test could be used above, no need for mock server here

File: crates/utils/src/http_client.rs [hunk 16]
```diff
@@ -0,0 +1,845 @@
...
+    }
+
+    /// # Edge Cases
+
+    /// Tests behavior with empty or problematic paths:
+    /// - Empty path segments
+    /// - Multiple slashes
+    /// - Unicode paths
+    #[test]
+    fn test_empty_path_handling() {
```
Comment (apoorvsadana) [inline crates/utils/src/http_client.rs:754]: let's have different cases for the same test case or in the same fn with multiple lines

File: crates/utils/src/http_client.rs [hunk 17]
```diff
@@ -0,0 +1,845 @@
...
+        assert!(request.body.is_some());
+        assert_eq!(request.body.unwrap().len(), (1024 * 1024) + 2);
+    }
+
+    /// Tests body serialization functionality:
+    /// - JSON serialization of different types
+    /// - Struct serialization
+    /// - Error handling for serialization failures
+    #[test]
+    fn test_request_builder_body_serialization() {
```
Comment (apoorvsadana) [inline crates/utils/src/http_client.rs:816]: let's move these edge cases above the integration ones?

File: crates/utils/src/http_client.rs (hunk 7 above)
Comment (Mohiiit) [reply crates/utils/src/http_client.rs]: done

File: .env.test (hunk 8 above)
Comment (Mohiiit) [reply .env.test]: done

File: crates/orchestrator/src/config.rs (hunk 9 above)
Comment (Mohiiit) [reply crates/orchestrator/src/config.rs]: done

File: crates/utils/src/http_client.rs (hunk 10 above)
Comment (Mohiiit) [reply crates/utils/src/http_client.rs]: changed

File: crates/utils/src/http_client.rs (hunk 11 above)
Comment (Mohiiit) [reply crates/utils/src/http_client.rs]: removed

File: crates/utils/src/http_client.rs (hunk 12 above)
Comment (Mohiiit) [reply crates/utils/src/http_client.rs:535]: added

File: crates/utils/src/http_client.rs (hunk 13 above)
Comment (Mohiiit) [reply crates/utils/src/http_client.rs]: deleted

File: crates/utils/src/http_client.rs (hunk 14 above)
Comment (Mohiiit) [reply crates/utils/src/http_client.rs]: removed

File: crates/utils/src/http_client.rs (hunk 16 above)
Comment (Mohiiit) [reply crates/utils/src/http_client.rs]: refactored

File: crates/utils/src/http_client.rs (hunk 17 above)
Comment (Mohiiit) [reply crates/utils/src/http_client.rs]: done

File: crates/utils/src/http_client.rs (hunk 15 above)
Comment (Mohiiit) [reply crates/utils/src/http_client.rs:557]: test moved and mock server removed as well

**Review by apoorvsadana — APPROVED** [review]

File: crates/prover-clients/atlantic-service/tests/lib.rs
Comment (Mohiiit) [reply crates/prover-clients/atlantic-service/tests/lib.rs:1]: changes it to make call to mock server

**Review by ocdbytes — APPROVED** [review]
Lgtm

File: crates/utils/src/http_client.rs [hunk 18]
```diff
@@ -0,0 +1,648 @@
...
+impl HttpClient {
+    /// Creates a new builder for constructing an HttpClient.
+    ///
+    /// # Arguments
+    /// * `base_url` - The base URL for all requests made through this client
+    ///
+    /// # Panics
+    /// Panics if the provided base URL is invalid
+    pub fn builder(base_url: &str) -> HttpClientBuilder {
+        println!("HttpClient builder created with base url: {:?}", base_url);
```
Comment (Mohiiit) [inline crates/utils/src/http_client.rs:86]: remove this

Comment (Mohiiit) [reply crates/utils/src/http_client.rs]: done

File: .env.example [hunk 19]
```diff
@@ -31,47 +30,48 @@ SQS_WORKER_TRIGGER_QUEUE_URL=
 ##### SNS #####
 
 ALERTS="sns"
-AWS_SNS_ARN="arn:aws:sns:us-east-1:000000000000:madara-orchestrator-arn"
+MADARA_ORCHESTRATOR_AWS_SNS_ARN="arn:aws:sns:us-east-1:000000000000:madara-orchestrator-arn"
 
 ##### DATABASE #####
 
 DATABASE=
```
Comment (apoorvsadana) [inline .env.example]: we had to actually delete `DATABASE`, `ALERTS` etc. i think these arne't used anymore

Comment (apoorvsadana) [reply .env.example]: basically, all env variables should have `MADARA_ORCHESTRATOR` prefix except AWS ones (which is already the case). ig the ones that don't have it can be deleted here
